### PR TITLE
Fix Snapshot State Machine Bug in Concurrent Clone and Delete

### PR DESCRIPTION
**Pull request by original-brownbear**
Unfortunately obvious bug when multiple deletes are queued up together with
a clone operation that was waiting for one of those deletes.

The test in this one isn't perfectly deterministic but very close to it (~80% failure rate). I'd like to get this merged asap as it's a bad bug and might be blocking other work, but I'd follow up with adding trickier test infrastructure to the mock repository that would allow a deterministic test here.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -2076,7 +2080,7 @@ private static boolean isWritingToRepository(SnapshotsInProgress.Entry entry) {
             // Entry is writing to the repo because it's finalizing on master
             return true;
         }
-        for (ObjectCursor<ShardSnapshotStatus> value : entry.shards().values()) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2079]: This completely failed to account for clones, so deletes incorrectly assumed that a clone wasn't writing and started concurrently.

Comment (tlrx) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:2079]: I've come across it multiple times and did not notice it... thanks.

**Review by tlrx — APPROVED** [review]
LGTM, thanks for fixing this Armin

**Comment by original-brownbear** [comment]
@elasticmachine update branch

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-1 (known issue)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-2 (known issue)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-2 (known issue)

**Comment by original-brownbear** [comment]
@elasticmachine update branch

**Comment by original-brownbear** [comment]
Thanks Tanguy!

**Pull request merged**